### PR TITLE
LG-2956: InlineCode component supports Polymorphic

### PR DESCRIPTION
**Pull request by brookescarlett**
## ✍️ Proposed changes

InlineCode component supports Polymorphic

🎟 _Jira ticket:_ [LG-2956](https://jira.mongodb.org/browse/LG-2956)

### For bug fixes, new features & breaking changes

- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

### For new components

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README

### CODE REVIEW

File: packages/typography/src/InlineCode/InlineCode.tsx [hunk 1]
```diff
@@ -20,26 +24,28 @@ import {
 } from './InlineCode.styles';
 import { InlineCodeProps } from './InlineCode.types';
 
-const InlineCode = React.forwardRef<HTMLElement, InlineCodeProps>(
+const InlineCode = InferredPolymorphic<InlineCodeProps>(
```
Comment (TheSonOfThomp) [inline packages/typography/src/InlineCode/InlineCode.tsx:27]: Does `InferredPolymorphic` need a default TS param?

Comment (TheSonOfThomp) [inline packages/typography/src/InlineCode/InlineCode.tsx:27]: Does InferredPolymorphic need a default TS param?

File: packages/typography/src/InlineCode/InlineCode.tsx [hunk 2]
```diff
@@ -60,14 +66,14 @@ const InlineCode = React.forwardRef<HTMLElement, InlineCodeProps>(
       </code>
     );
 
-    if (isAnchor) {
+    if (needsWrapper) {
```
Comment (TheSonOfThomp) [inline packages/typography/src/InlineCode/InlineCode.tsx:69]: Don't love the conditional rendering here. I think we're trying to move away from `renderedXYZ` type variables

Comment (TheSonOfThomp) [reply packages/typography/src/InlineCode/InlineCode.tsx:69]: I would recommend something like:
```tsx 
  const { Component } = useInferredPolymorphic(as, rest, Fragment);
  const isFragment = isUndefined(as) // or isEqual(Component, Fragment)
  ...
  <Component {...(isFragment && {
      className: cx(...),
      ...rest
    }
  )}>
    <code>
      ...
    </code>
  </Component>

```

File: packages/typography/src/InlineCode/InlineCode.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        const utils = render(<InlineCode href="string">{children}</InlineCode>);
+        const component = utils.container.firstChild;
+        expect((component as HTMLElement).tagName.toLowerCase()).toBe('a');
+      });
+
+      test('it renders as component passed to "as" prop when set', () => {
+        const utils = render(<InlineCode as="div">{children}</InlineCode>);
+        const component = utils.container.firstChild;
+        expect((component as HTMLElement).tagName.toLowerCase()).toBe('div');
+      });
```
Comment (TheSonOfThomp) [inline packages/typography/src/InlineCode/InlineCode.spec.tsx:27]: Should also verify that there's still a `<code>` element in the DOM

File: packages/typography/src/InlineCode/InlineCode.tsx (hunk 1 above)
Comment (brookescarlett) [reply packages/typography/src/InlineCode/InlineCode.tsx:27]: 👍

File: packages/typography/src/InlineCode/InlineCode.tsx (hunk 2 above)
Comment (brookescarlett) [reply packages/typography/src/InlineCode/InlineCode.tsx:69]: Fragments don't accept props though

Comment (TheSonOfThomp) [reply packages/typography/src/InlineCode/InlineCode.tsx:69]: Oops I had the check backwards. You'd need to check `!isFragment && ...`. It's a bit of an ugly solution tbf

**Review by TheSonOfThomp — APPROVED** [review]

**Pull request merged**